addons/portal/controllers/thread.py

```diff
@@ -0,0 +1,20 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.http import request
+from odoo.addons.mail.controllers import thread
+from odoo.addons.portal.utils import validate_thread_with_hash_pid, validate_thread_with_token
+
+
+class ThreadController(thread.ThreadController):
+
+    def _prepare_post_data(self, post_data, thread, **kwargs):
+        post_data = super()._prepare_post_data(post_data, thread, **kwargs)
+        if request.env.user._is_public():
+            if validate_thread_with_hash_pid(thread, kwargs.get("hash"), kwargs.get("pid")):
+                post_data["author_id"] = int(kwargs["pid"])
+            elif (
+                (partner := next(iter(thread._mail_get_partners().values())))
```

Comment:
Can't this raise `StopIteration`? Shouldn't we catch and ignore it (or provide default empty)?

Also `.values()` can probably be replaced by `[thread.id]`?

Reply:
More than that it doesn't actually work, it just gets the first dict value, not the first partner.

I'll push the fix.